crucible-llvm/src/Lang/Crucible/LLVM/ArraySizeProfile.hs

```diff
@@ -0,0 +1,153 @@
+------------------------------------------------------------------------
+-- |
+-- Module           : Lang.Crucible.LLVM.ArraySizeProfile
+-- Description      : Execution feature to observe argument buffer sizes
+-- Copyright        : (c) Galois, Inc 2019
+-- License          : BSD3
+-- Maintainer       : Andrei Stefanescu <andrei@galois.com>
+-- Stability        : provisional
+------------------------------------------------------------------------
+
+{-# OPTIONS -Wall -Werror #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Lang.Crucible.LLVM.ArraySizeProfile
+  ( Profile
+  , arraySizeProfile
+  ) where
+
+import qualified Control.Lens as Lens
+
+import Data.Type.Equality
+import Data.Foldable
+import Data.IORef
+import Data.Text (Text)
+import qualified Data.Text as Text
+import qualified Data.Vector as Vector
+
+import Data.Parameterized.SymbolRepr
+import Data.Parameterized.Context
+import Data.Parameterized.TraversableFC
+
+import qualified Lang.Crucible.Backend as C
+import qualified Lang.Crucible.Types as C
+import qualified Lang.Crucible.CFG.Core as C
+import qualified Lang.Crucible.Simulator.BoundedExec as C
+import qualified Lang.Crucible.Simulator.CallFrame as C
+import qualified Lang.Crucible.Simulator.EvalStmt as C
+import qualified Lang.Crucible.Simulator.ExecutionTree as C
+import qualified Lang.Crucible.Simulator.GlobalState as C
+import qualified Lang.Crucible.Simulator.Intrinsics as C
+import qualified Lang.Crucible.Simulator.RegMap as C
+
+import qualified Lang.Crucible.LLVM.Extension as C
+import qualified Lang.Crucible.LLVM.Intrinsics.Common as C
+import qualified Lang.Crucible.LLVM.MemModel as C
+
+import qualified Lang.Crucible.LLVM.MemModel.Generic as G
+
+import qualified What4.Interface as W4
+
+type Profile = (Text, [[Maybe Int]])
+
+ptrStartsAlloc ::
+  W4.IsExpr (W4.SymExpr sym) =>
+  C.LLVMPtr sym w ->
+  Maybe ()
+ptrStartsAlloc (C.llvmPointerView -> (_, W4.asUnsignedBV -> Just 0)) = Just ()
+ptrStartsAlloc _ = Nothing
+
+ptrAllocSize ::
+  forall sym w. W4.IsExpr (W4.SymExpr sym) =>
+  [G.MemAlloc sym] ->
+  C.LLVMPtr sym w ->
+  Maybe Int
+ptrAllocSize mem (C.llvmPointerView -> (blk, _)) = msum $ inAlloc <$> mem
+  where inAlloc :: G.MemAlloc sym -> Maybe Int
+        inAlloc (G.Alloc _ _ Nothing _ _ _) = Nothing
```

Comment:
I prefer
```
inAlloc mem_alloc
  | (G.Alloc _ a (Just sz) _ _ _) <- mem_alloc
  , Just a ==  W4.asNat blk ->
    Just $ fromIntegral $ W4.asUnsignedBV sz
  | otherwise -> Nothing
```